birdvoxpaint/util.py

```diff
@@ -0,0 +1,225 @@
+        block_n_frames = sample2frame(block_length, frame_length, hop_length)
+        y_blocks = librosa.stream(filename,
+            block_length=block_n_frames,
+            frame_length=frame_length,
+            hop_length=hop_length)
+
+        def maybe_resample(y_blocks):
+            # will throw an error if audio is not valid
+            # resample if we have a different sr
+            for y in y_blocks:
+                librosa.util.valid_audio(y, mono=True)
+                if sr != orig_sr:
+                    y = librosa.resample(y, orig_sr, sr)
+                yield y
+
+        y_blocks = maybe_resample(y_blocks)
+
+    else:
+        if y is None or sr is None:
+            raise ParameterError(
+                'At least one of (y, sr) or filename must be provided')
+
+        librosa.util.valid_audio(y, mono=True)
+
+        # get block length, remove frame overhang (reuses conversions which does flooring)
+        block_length = max(segment_duration * sr, frame_length) * n_blocks # min block size = 1 frame
+        block_length = sample2frame(block_length, frame_length, hop_length) # drop overhang
+        block_length = frame2sample(block_length, frame_length, hop_length) # convert back
+
+        # get frames from array
+        y_blocks = librosa.util.frame(y, block_length, block_length).T
+
+    if full_frames: # drop any frames that are incomplete
+        y_blocks = (y for y in y_blocks if y.size == block_length)
+
+    return y_blocks, sr
+
+
+# def multifile_stream(filenames, block_length):
+#     '''Chain multiple file audio block streams into a single stream. It uses samples
+#     from the next file to complete a previously incomplete block
+#     '''
+#     raise NotImplemented("I haven't really played around or tested this. I just drafted it.")
+#     remaining = None
+#     for filename in filenames:
+#         offset = 0 # TODO: uhhh I guess ? - idk
+#         # TODO: Load up to block_length -
+#         if remaining is not None:
+#             offset = block_length - len(remaining)
+#             y_complete, _ = librosa.load(filename, duration=offset)
+#             remaining = np.concatenate([remaining, y_complete])
+#
+#             if len(remaining) == block_length: # there were enough samples in the new file
+#                 yield remaining
+#                 remaining = None # clear
+#             else: # super short file ??? - better to be robust to all types of weather.
+#                 continue
+#
+#         y_blocks = librosa.stream(filename,
+#             offset=offset,
+#             block_length=block_n_frames,
+#             frame_length=frame_length,
+#             hop_length=hop_length)
+#
+#         for y in y_blocks:
+#             librosa.util.valid_audio(y, mono=True)
+#             if sr != orig_sr:
+#                 librosa.resample(y, orig_sr, sr)
+#             yield y
+
+
+
+def spec_frame(S, segment_length):
```

Comment:
let's just use cutting-edge `librosa.util.frame`
https://github.com/librosa/librosa/pull/966
v0.7.1 will land in ~10 days or so

Reply:
Oooooo I've been waiting for this